buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionArchiveSetupPlugin.java

```diff
@@ -0,0 +1,196 @@
+                : new DistributionArchive(project.getTasks().register(name, Zip.class), explodedDist, name);
+        });
+        // Each defined distribution archive is linked to a subproject.
+        // A distribution archive definition not matching a sub project will result in build failure.
+        container.whenObjectAdded(distributionArchive -> {
+            var subProjectName = buildTaskToSubprojectName(distributionArchive.getArchiveTask().getName());
+            project.project(subProjectName, sub -> {
+                sub.getPlugins().apply("base");
+                sub.getArtifacts().add("default", distributionArchive.getArchiveTask());
+                var explodedArchiveTask = distributionArchive.getExplodedArchiveTask();
+                var defaultConfiguration = sub.getConfigurations().getByName("default");
+                var publications = defaultConfiguration.getOutgoing();
+                var variant = (ConfigurationVariantInternal) publications.getVariants().maybeCreate("directory");
+                variant.getAttributes().attribute(ArtifactAttributes.ARTIFACT_FORMAT, ArtifactTypeDefinition.DIRECTORY_TYPE);
+                variant.artifactsProvider(() -> Collections.singletonList(new DirectoryPublishArtifact(explodedArchiveTask)));
+            });
+        });
+        project.getExtensions().add("distribution_archives", container);
+    }
+
+    private void configureGeneralTaskDefaults(Project project) {
+        // common config across all copy / archive tasks
+        project.getTasks().withType(AbstractCopyTask.class).configureEach(t -> {
+            t.dependsOn(project.getTasks().withType(EmptyDirTask.class));
+            t.setIncludeEmptyDirs(true);
+            t.setDirMode(0755);
+            t.setFileMode(0644);
+        });
+
+        // common config across all archives
+        project.getTasks().withType(AbstractArchiveTask.class).configureEach(t -> {
+            String subdir = buildTaskToSubprojectName(t.getName());
+            t.getDestinationDirectory().set(project.file(subdir + "/build/distributions"));
+            t.getArchiveBaseName().set(subdir.contains("oss") ? "elasticsearch-oss" : "elasticsearch");
+        });
+    }
+
+    private void configureTarDefaults(Project project) {
+        // common config across all tars
+        project.getTasks().withType(SymbolicLinkPreservingTar.class).configureEach(t -> {
+            t.getArchiveExtension().set("tar.gz");
+            t.setCompression(Compression.GZIP);
+        });
+    }
+
+    private void registerEmptyDirectoryTasks(Project project) {
+        // CopySpec does not make it easy to create an empty directory so we
+        // create the directory that we want, and then point CopySpec to its
+        // parent to copy to the root of the distribution
+        File logsDir = new File(project.getBuildDir(), "logs-hack/logs");
+        project.getExtensions().add("logsDir", new File(project.getBuildDir(), "logs-hack/logs"));
```

Comment:
Shouldn't these be extra properties instead of extensions? I assume there's not too much difference but extensions end up get decorated by Gradle at runtime which we probably don't need in this case.

Reply:
good point. fixed.